### PR TITLE
Better error messages

### PR DESCRIPTION
**Pull request by henrinormak**
Goal was to improve the legibility of error messages, with emphasis on nested complex types.

Some examples of errors

```
import { Array, String, Number, Boolean, Record } from 'runtypes';
const Flags = Array(Boolean);
Flags.check([true, true, 's']);
// { Error: Expected boolean but was string, key: '[2]' }

const Page = Record({
  title: String
});

const Person = Record({
  name: String,
  age: Number,
  likes: Array(Page)
});
const People = Array(Person);
People.check([{ name: 'Jack', likes: [] }]);
// { Error: Expected age to be number, key: '[0].age' }

People.check([{ name: 'Jack', age: 10, likes: [{ name: undefined, likes: [] }] }]);
// { Error: Expected title to be string, key: '[0].likes.[0].title' }
```

~*Note:* Before merging, want to add some error specific tests (currently none of the tests checks the errors).~

### CODE REVIEW

**Comment by henrinormak** [comment]
Also improved union types, which now report the expected value along with the value that was checked.

```
const Pet = Record({
  name: String
});

const PersonOrPet = Union(
  Person,
  Pet
);

PersonOrPet.check({ foo: 'bar' });
// Error: None of the alternatives { name: string; age: number; likes: { title: string; }[]; },{ name: string; } matched {"foo":"bar"}
```

**Comment by henrinormak** [comment]
Finally, also improved errors for Partial type (including if the field is itself a complex type, such as a union):

```
const PartialPerson = Partial({
  name: String,
  age: Union(Number, String),
  likes: Array(Page)
});

PartialPerson.check({ name: 'Jack', age: { value: 10 } });
// { Error: Expected (number | string) | undefined, got {"value":10} key: 'age' }
```

File: src/types/dictionary.ts [hunk 1]
```diff
@@ -36,7 +42,12 @@ export function Dictionary<V extends Runtype>(value: V, key = 'string'): any {
           if (isNaN(+k))
             throw validationError(`Expected dictionary key to be a number but was string`);
         }
-        value.check((x as any)[k]);
+
+        try {
```
Comment (erkiesken) [inline src/types/dictionary.ts]: I wonder what's the performance impact like for creating all these extra try/catch contexts in this loop?

File: src/types/array.ts [hunk 2]
```diff
@@ -12,7 +12,15 @@ function Arr<E extends Runtype>(element: E): Arr<E> {
   return create<Arr<E>>(
     xs => {
       if (!Array.isArray(xs)) throw validationError(`Expected array but was ${typeof xs}`);
-      for (const x of xs) element.check(x);
+
+      for (const x of xs) {
+        try {
```
Comment (erkiesken) [inline src/types/array.ts]: Ditto here on array element loop using separate try/catch inside. Maybe use `benchmark` lib to measure if moving try/catch outside the loop make a noticable difference?

Comment (henrinormak) [reply src/types/array.ts]: Good suggestion, will do that

Comment (erkiesken) [reply src/types/array.ts]: Try it with some simple-ish arrays and also some huge ones (1000+) items. Same for dicts.

**Comment by henrinormak** [comment]
Did some benchmarking, sort of as expected, when there is a fault, the new code is slower, with correct codepaths, the change remains within the standard deviation of the old error handling code (albeit it is slower). I used a large array (10k items) and a small one (10 items) of random data that checked against the Record type. Similarly with dictionaries.

```
Current production code - 8 tests completed.

  Correct array (large)      x     1,241 ops/sec ±10.85% (76 runs sampled)
  Correct dictionary (large) x       440 ops/sec ±7.77% (78 runs sampled)
  Faulty array (large)       x     1,374 ops/sec ±9.20% (83 runs sampled)
  Faulty dictionary (large)  x       445 ops/sec ±7.98% (81 runs sampled)
  Correct array (small)      x 1,412,369 ops/sec ±10.79% (80 runs sampled)
  Correct dictionary (small) x   401,691 ops/sec ±7.86% (50 runs sampled)
  Faulty array (small)       x    59,692 ops/sec ±7.24% (81 runs sampled)
  Faulty dictionary (small)  x    53,472 ops/sec ±6.55% (82 runs sampled)

With the new errors - 8 tests completed.

  Correct array (large)      x     1,465 ops/sec ±8.97% (83 runs sampled)
  Correct dictionary (large) x       432 ops/sec ±8.28% (78 runs sampled)
  Faulty array (large)       x     1,333 ops/sec ±10.12% (80 runs sampled)
  Faulty dictionary (large)  x       455 ops/sec ±4.93% (81 runs sampled)
  Correct array (small)      x 1,353,563 ops/sec ±9.58% (84 runs sampled)
  Correct dictionary (small) x   388,521 ops/sec ±9.10% (47 runs sampled)
  Faulty array (small)       x    41,303 ops/sec ±5.99% (81 runs sampled)
  Faulty dictionary (small)  x    35,289 ops/sec ±8.72% (76 runs sampled)
```
